examples_tests/31.SkinningDataBenchmark/main.cpp

```diff
@@ -348,133 +349,160 @@ int main()
-                boneMatricesRows[s3pc.matrixOffsets[3] + i] = boneMatrices[i].getRow(3);
-
-                normalMatricesRows[s3pc.matrixOffsets[0] + i] = core::matrix4SIMD(normalMatrices[i]).getRow(0);
-                normalMatricesRows[s3pc.matrixOffsets[1] + i] = core::matrix4SIMD(normalMatrices[i]).getRow(1);
-                normalMatricesRows[s3pc.matrixOffsets[2] + i] = core::matrix4SIMD(normalMatrices[i]).getRow(2);
-            }
-
             const size_t boneMatricesRowsByteSize = sizeof(core::vectorSIMDf) * boneMatricesRows.size();
             const size_t normalMatricesRowsByteSize = sizeof(core::vectorSIMDf) * normalMatricesRows.size();
             const size_t normalMatricesOffset = sizeof(core::vectorSIMDf) * BONE_VEC_MAX_CNT;
 
             driver->updateBufferRangeViaStagingBuffer(drawDataBuffer[2].get(), 0u, boneMatricesRowsByteSize, boneMatricesRows.data());
             driver->updateBufferRangeViaStagingBuffer(drawDataBuffer[2].get(), normalMatricesOffset, normalMatricesRowsByteSize, normalMatricesRows.data());
-            driver->pushConstants(gpuPipelineLayout[2].get(), asset::ISpecializedShader::ESS_VERTEX, 0u, sizeof(Shader3PushConstants), &s3pc);
         }
         break;
-        case components:
+        case 3:
         {
-            for (uint32_t i = 0u; i < diskCount; i++)
-            {
-                for (uint32_t j = 0u; j < 16u; j++)
-                    boneMatricesComponents[s4pc.matrixOffsets[j] + i] = *(boneMatrices[i].pointer() + j);
-
-                for (uint32_t j = 0u; j < 9u; j++)
-                    normalMatricesComponents[s4pc.matrixOffsets[j] + i] = *(boneMatrices[i].pointer() + j); //fix!
-            }
-
             const size_t boneMatricesCompByteSize = sizeof(float) * boneMatricesComponents.size();
             const size_t normalMatricesCompByteSize = sizeof(float) * normalMatricesComponents.size();
 
             driver->updateBufferRangeViaStagingBuffer(drawDataBuffer[3].get(), 0u, boneMatricesCompByteSize, boneMatricesComponents.data());
             driver->updateBufferRangeViaStagingBuffer(drawDataBuffer[3].get(), boneMatricesCompByteSize, normalMatricesCompByteSize, normalMatricesComponents.data());
-            driver->pushConstants(gpuPipelineLayout[3].get(), asset::ISpecializedShader::ESS_VERTEX, 0u, sizeof(Shader4PushConstants), &s4pc);
         }
         break;
         default:
             assert(false);
         }
+    };
 
-        driver->drawIndexedIndirect(mdi.vtxBindings, mdi.mode, mdi.indexType, mdi.indexBuff.get(), mdi.indirectDrawBuff.get(), mdi.offset, mdi.maxCount, mdi.stride);
-        
-        driver->endScene();
+    //run benchmark for all cases
+    constexpr uint32_t iterationCnt = 10000u;
+
+    COpenGLDriver* driverOGL = dynamic_cast<COpenGLDriver*>(driver);
+    IQueryObject* query = driver->createElapsedTimeQuery();
+
+    for (uint32_t caseID = 0u; caseID < 4u; caseID++)
```

Comment:
you need one `driver->beginScene` at the start and then set the correct FBO (depth only, really low resolution)

Reply:
and clear the Z-Buffer to the near value (1.f)

Reply:
all that for the benchmark mode obviously.